src/libraries/Microsoft.Extensions.Logging.Console/src/StringWriter.cs

```diff
@@ -0,0 +1,182 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.IO;
+using System.Text;
+using System.Collections.Concurrent;
+
+namespace Microsoft.Extensions.Logging.Console
+{
+    internal class StringWriter : TextWriter
+    {
+        private readonly StringBuilder _stringBuilder;
+        internal bool DisableColors { get; set; }
+
+        public override Encoding Encoding => Encoding.Unicode;
+
+        public StringWriter(bool disableColors = false)
+        {
+            _stringBuilder = new StringBuilder();
+            DisableColors = disableColors;
+        }
+
+        public override void Write(char value)
+        {
+            _stringBuilder.Append(value);
+        }
+
+        public bool SetBackgroundColor(ConsoleColor? background)
+        {
+            if (DisableColors || background == _colorSettings.bg)
+            {
+                return false;
+            }
+            _colorSettings.bg = background;
+            if (background.HasValue)
+            {
+                Write(GetBackgroundColorEscapeCode(background.Value));
+            }
+            else
+            {
+                Write(DefaultBackgroundColor);
+            }
+            return true;
+        }
+
+        public bool SetForegroundColor(ConsoleColor? foreground)
+        {
+            if (DisableColors || foreground == _colorSettings.fg)
+            {
+                return false;
+            }
+            _colorSettings.fg = foreground;
+            if (foreground.HasValue)
+            {
+                Write(GetForegroundColorEscapeCode(foreground.Value));
+            }
+            else
+            {
+                Write(DefaultForegroundColor);
+            }
+            return true;
+        }
+
+        private (ConsoleColor? bg, ConsoleColor? fg) _colorSettings = (null, null);
+
+        public void ResetColor()
+        {
+            SetBackgroundColor(null);
+            SetForegroundColor(null);
+        }
+
+        private const string DefaultForegroundColor = "\x1B[39m\x1B[22m"; // reset to default foreground color
+        private const string DefaultBackgroundColor = "\x1B[49m"; // reset to the background color
```

Comment:
these should go to AnsiLogConsole.cs file

Reply:
Removed duplicate code in AnsiLogConsole.cs

AnsiLogConsole.Write(string message) does not need to do any ansi color processing anymore. the message it gets already contains the color codes in it, provided by the Formatter.

As @eerhardt recommended I am now using an extension method in the SimpleConsoleFormatter to embed color codes in the message, (only does that when DisableColors is false).